### ISSUE TITLE
reset indexes before writing to parquet to save space

### ISSUE THREAD

**Issue by bmschmidt** [massivetexts/htrc-feature-reader#8]
Changing the structure of the parquet files may be able to really improve the compression rates by about a 30% reduction in size. 

I thought that maybe sorting by a different column might let the RLE in parquet kick in. That seems not to matter; but what *does* really help is resetting indexes before writing. Having a single index reduces file size a lot; having double indices as in the example does less, and whatever hierarchical format pandas returns after `.count()` seems to be a lot bigger than necessary.

For me, simply rewriting the test file out again produces a higher size than in the repo--not sure what's going on there.

But see this list of different formats. in a single directory from largest to smallest.

The best performing strategy is this:

```python
source_file = '../data/parquet/mdp.39015028036104.tokens.parquet'
d = pd.read_parquet(source_file)
psorted = d.reset_index().sort_values(['section', 'pos', 'count', 'token'])
psorted = psorted.set_index(['page'])
psorted.to_parquet(source_file + '.page_index_spct_sort', compression = 'snappy')
```

All file sizes

```
-rw-r--r--@ 1 bschmidt  staff  865958 Dec 29 23:53 mdp.39015028036104.tokens.parquet.no_index
-rw-r--r--@ 1 bschmidt  staff  389785 Dec 29 23:53 mdp.39015028036104.tokens.csv.gz
-rw-r--r--@ 1 bschmidt  staff  371221 Dec 29 23:53 mdp.39015028036104.tokens.parquet.rewritten
-rw-r--r--@ 1 bschmidt  staff  369581 Dec 29 23:28 mdp.39015028036104.tokens.parquet
-rw-r--r--@ 1 bschmidt  staff  340589 Dec 29 23:53 mdp.39015028036104.tokens.parquet.pos_index_spct_sort
-rw-r--r--@ 1 bschmidt  staff  296226 Dec 29 23:53 mdp.39015028036104.tokens.parquet.pos_index_ct_sort
-rw-r--r--@ 1 bschmidt  staff  281785 Dec 29 23:53 mdp.39015028036104.tokens.parquet.tpc_sort
-rw-r--r--@ 1 bschmidt  staff  276082 Dec 29 23:53 mdp.39015028036104.tokens.parquet.cpt_sorted
-rw-r--r--@ 1 bschmidt  staff  271149 Dec 29 23:53 mdp.39015028036104.tokens.parquet.ptc_sort
-rw-r--r--@ 1 bschmidt  staff  268107 Dec 29 23:53 mdp.39015028036104.tokens.parquet.pct_sort
-rw-r--r--@ 1 bschmidt  staff  266798 Dec 29 23:53 mdp.39015028036104.tokens.parquet.page_index_spct_sort
```

Full code and description of each sorting strategy, except for the csv-based ones.

```python
source_file = '../data/parquet/mdp.39015028036104.tokens.parquet'
d = pd.read_parquet(source_file)

d.to_parquet(source_file + '.rewritten', compression='snappy')

resorted = d.reset_index().sort_values(['count', 'pos', 'token'])
resorted = resorted.set_index(['page', 'section'])
resorted.to_parquet(source_file + '.cpt_sorted', compression = 'snappy')

no_index = d.reset_index().sort_values(['section', 'pos', 'page', 'count', 'token'])
no_index.to_parquet(source_file + '.no_index', compression = 'snappy')

tsorted = d.reset_index().sort_values(['token', 'pos', 'count'])
tsorted = tsorted.set_index(['page', 'section'])
tsorted.to_parquet(source_file + '.tpc_sort', compression = 'snappy')

psorted = d.reset_index().sort_values(['pos', 'token', 'count'])
psorted = psorted.set_index(['page', 'section'])
psorted.to_parquet(source_file + '.ptc_sort', compression = 'snappy')

psorted = d.reset_index().sort_values(['pos', 'token', 'count'])
psorted = psorted.set_index(['page', 'section'])
psorted.to_parquet(source_file + '.ptc_sort', compression = 'snappy')

psorted = d.reset_index().sort_values(['pos', 'count', 'token'])
psorted = psorted.set_index(['page', 'section'])
psorted.to_parquet(source_file + '.pct_sort', compression = 'snappy')

psorted = d.reset_index().sort_values(['count', 'token'])
psorted = psorted.set_index(['page', 'section', 'pos'])
psorted.to_parquet(source_file + '.pos_index_ct_sort', compression = 'snappy')

psorted = d.reset_index().sort_values(['section', 'pos', 'count', 'token'])
psorted = psorted.set_index(['page'])
psorted.to_parquet(source_file + '.page_index_spct_sort', compression = 'snappy')
```

**Comment by organisciak** [comment]
Which engine are you using for parquet? I'm not seeing any changes from rewriting, and the size difference between fastparquet and pyarrow is much larger than you're seeing.

None of the sortings make much practical sense, but some people may prefer the lower space versus faster performance. Do you think we should add a 'dense_sort' export argument? Of course, at that point, maybe 'snappy' compression is also an ill-fit.

I expanded this to test size on the 15 files in `https://github.com/massivetexts/htrc-feature-reader/tree/master/data/PZ-volumes`. What I find interesting:

- Sorting the index versus sorting by column doesn't seem to make any difference on my system. e.g. `df.reorder_levels([...]).sort_index()`

- I usually don't save all the information, particularly dropping the `section`. However, the sorted full versions are practically the same size as those without `section` and `pos` columns. That's notable because it's not just fewer columns, but fewer rows. Here are the best permutations of different column selections, for the 15 books:

```
token,page Size: 2.53M
section,token,page Size: 2.54M
pos,token,page Size: 2.66M
pos,section,token,page Size: 2.66M
```

- The placement of `section` in the sort order is negligible. Consider these sort orders:

```
section,pos,token,page Size: 2.65M
pos,section,token,page Size: 2.66M
pos,token,section,page Size: 2.67M
pos,token,page,section Size: 2.67M
```

- The data-oriented ordering of page, section,token,pos is the *largest* permutation. Dropping `section` for brevity here:

```
page,token,pos Size: 3.35M
page,pos,token Size: 2.95M
token,page,pos Size: 2.83M
token,pos,page Size: 2.76M
pos,page,token Size: 2.96M
pos,token,page Size: 2.66M
```

- Perhaps confirming the obvious, but if you only keep the 'body' (effectively making section a useless column with only one possible value), the column pretty much doesn't take up space. 

```
Just body: page,token Size: 2.64M
Just body: page,section,token Size: 2.65M
Just body: page,token,section Size: 2.65M
Just body: section,page,token Size: 2.65M

Just body: token,page Size: 2.48M
Just body: section,token,page Size: 2.49M
Just body: token,page,section Size: 2.49M
Just body: token,section,page Size: 2.49M
```

- Also expected, the space saved from saving lowercased tables is identical with all sortings. e.g.

```
page,lowercase,pos Size: 3.15M
page,token,pos Size: 3.35M

pos,lowercase,page Size: 2.50M
pos,token,page Size: 2.66M
```

**Comment by bmschmidt** [comment]
This is really weird. What an odd format. I would have thought there'd be a little more consistency between them. I think I'm using `pyarrow`, but hard to tell. Share the code you used for this? I would broaden the parameter space to include:

1. Ordering
2. Which fields get indexes
3. Fastparquet vs pyarrow.

## What to cache

> I usually don't save all the information, particularly dropping the section. However, the sorted full versions are practically the same size as those without section and pos columns. 

I get the impression that it would be best for this package to only ever cache one parquet format; the 'page,section,pos,token,count' one. Anything else might cause confusion, even for you and me. Especially since parquet can read in only the columns needed for a particular operation, which reduces the overhead further for storing extra columns. If you want to write out something like chunked counts, just do it manually rather than bundle it in the library; it should take barely any more code.

## Time-space tradeoffs

> but some people may prefer the lower space versus faster performance. Do you think we should add a 'dense_sort' export argument? Of course, at that point, maybe 'snappy' compression is also an ill-fit.

Are you sure there's a tradeoff? Using a simple test of token counts and timeit, I seem to get much shorter times on the shorter files. 40ms for the page_index_spct_sort, 80ms for the current version. The difference might be larger on non-SSD media where reads are costlier.

```python
%%timeit

fin = "../data/parquet/mdp.39015028036104.tokens.parquet.page_index_spct_sort"
_ = pd.read_parquet(fin, columns=['token', 'count'])\
.reset_index().groupby(["token"])['count'].sum()
```

If we insist that there must be hierarchical indexes with groups like in the package on the returned value, things might be different, though. Index resets seem to be cheap: ~9ms.

**Comment by bmschmidt** [comment]
I realized I don't understand what this means; "Sorting the index versus sorting by column doesn't seem to make any difference on my system. e.g. df.reorder_levels([...]).sort_index()." Could you elaborate a little?

**Comment by organisciak** [comment]
## Time-space payoff

It's quicker to read the input file, but that's balanced by the extra work to then sort it back to a proper hierarchical index. It's not a bad performance hit - about 0.053s/volume or 14.8hrs/1million.

I like that balance, though want to make it optional since I developed `save_parquet` foremost to offload computation to pre-processing of the EF and speed up downstream use of the dataset.

```
page,token,pos Size: 3.35M Reading time: 0.34s
page,pos,token Size: 2.95M Reading+Sorting time: 1.14s
token,page,pos Size: 2.83M Reading+Sorting time: 1.09s
token,pos,page Size: 2.76M Reading+Sorting time: 1.09s
pos,page,token Size: 2.96M Reading+Sorting time: 1.13s
pos,token,page Size: 2.66M Reading+Sorting time: 1.13s
```

## Reset Index

I was saying the `reset_index` isn't necessary for space. You can do

```
df = df.reset_index().sort_values(permutation).set_index(permutation[0])
```

or: 

```
df.reorder_levels(perm).sort_index()
```

and the output files will be identical sizes. For performance, the latter is notable quicker. e.g.

```
reset_index 	 page,section,token,pos Size: 3.37M Time: 0.3922436237335205
reorder_levels 	 page,section,token,pos Size: 3.37M Time: 1.8180656433105469
```

### Output format

Moving away from a rigidly expected format has been the most significant internal refactor in years, alongside the abstraction of the parsing from the Volume class, so I don't expect to roll it back. Not all EF-like datasets will need pages or pos or sections. Currently, the default is to save all columns. The data can only be written out in a format that's both self-documenting if manually opened and machine-readable through the library.

### Notebook

https://nbviewer.jupyter.org/github/massivetexts/htrc-feature-reader/blob/ee5824dd7598b2cf45c4da45e02ad55ca3b2066b/dev/ParquetSizeComparisons.ipynb

**Comment by bmschmidt** [comment]
# Hierarchical indexes

I've never used the hierarchical indexes in this package, TBH. My typical code looks like:

```python
tokenlist.reset_index().groupby(["page", "token"]).agg("sum")
```

Given that index creation appears to actually be somewhat expensive, I'd love to be able to turn them off and get just unindexed pandas frames from various different functions.

# Indexless storage is good, after all.

Turns out the reason that my indexless storage was taking up so much space because pandas was writing index values it made up anyway to preserve the original sort order. (How the integer indices managed to be the size of the rest of files kinda baffles me, though--I guess just uncompressed uint64 s?). This removes a bit of my confusion. It also makes it possible to include 'count' in the sortlist, which gets the size down even farther: 2.53 MB, vs 2.68M for the previous leader.

Creation code:
```python
df.reset_index(inplace = True)
df.sort_values([ "pos", "count", "token", "page"], inplace = True)
df.to_parquet(outpath, engine='pyarrow', index = False)
```

It also seems to be substantially faster to read than anything but the pre-created index, presumably because the old index doesn't have to be destroyed and so we can use one of your different strategies more effectively (sort first, then build index).

```python
df = pd.read_parquet(outpath)
 df.sort_values(preferred_order, inplace = True)
 df.set_index(preferred_order, inplace = True)
```

The performance is quite good compared to the other methods:


```
token,pos,pagenoindex Size: 2.53M Reading+Sorting time: 0.99s`
```


```
page,token,pos Size: 3.37M Reading time: 0.55s
page,pos,token Size: 2.97M Reading+Sorting time: 1.67s
token,page,pos Size: 2.85M Reading+Sorting time: 1.74s
token,pos,page Size: 2.78M Reading+Sorting time: 1.62s
pos,page,token Size: 2.98M Reading+Sorting time: 1.70s
pos,token,page Size: 2.68M Reading+Sorting time: 1.72s
```


Full code:

```python
totaltime = 0

for path in ogpaths:
    df = pd.read_parquet(path).reset_index().groupby(list(perm))[['count']].sum().reset_index()
    df = df.sort_values([ "pos", "count", "token", "page"])
    
    outpath = os.path.join(rewritedir.name, os.path.split(path)[1])
    
    df.to_parquet(outpath, engine='pyarrow', index = False)
    # Measure read and sort-if-necessary times
    starttime = time.time()
    df = pd.read_parquet(outpath)
    if df.index.names != preferred_order:
        # fastest
        #df = df.reorder_levels(preferred_order).sort_index()
        # alt
        #df.set_index(preferred_order).sort_index(inplace = True)
        # potentially faster, but not sure yet if there's a downstream performance hit
        df.sort_values(preferred_order, inplace = True)
        df.set_index(preferred_order, inplace = True)
    totaltime += (time.time() - starttime)

if list(perm) == preferred_order:
    print(statdir(",".join(perm), rewritedir.name), "Reading time: {:.2f}s".format(totaltime))
else:
    print(statdir(",".join(perm), rewritedir.name), "Reading+Sorting time: {:.2f}s".format(totaltime))
```

**Comment by bmschmidt** [comment]
I think the thing blocking this issue is how important Ids are. I think in general they're a user convenience, but take away some performance. So I see two options here:

1. Store sorted unindexed parquet, and add an argument, by default false, to the core Volume class that allows it return unindexed data. This would be valuable in high-performance settings; I remember when I ran the SRP features, I wrote my own JSON parsing code because it turned out to be faster than this module, I think now probably because of index creation.

2. Store fully indexed parquet on disk (maybe w/ some sorting, still, on count?) and don't add that argument.

I prefer the first, but it might be tricky.

**Comment by organisciak** [comment]
I think I've lost the thread on this conversation and the motivations here, but I think I'm wrapping my head around it. The action items here are:

- [ ] Add a save_parquet argument to allow you to save unindexed and densely albeit unintuitively sorted parquet files. This makes sense for certain cases, and an extra argument option on the method to support those cases is great. The new code has an option for resetting the index, so all that's needed is sorting.
- [ ] Add a check for columns on parquet reading, to allow non-standard formats to be read. For example, if it looks like an unindexed file, sort and set index.

**Comment by organisciak** [comment]
Oh, and 
- [ ] passing through an unindexed file on read (i.e. turn off the check for format). This might be allow for some issues internally, but I propose we assume that somebody doing this knows what they're doing and straightforwardly give it to them.

**Comment by bmschmidt** [comment]
Introspecting, I think the motivation here is really my general dislike of pandas indexes as much as the specific savings here. Plus curiosity about Parquet. I haven't tested it yet, but I think allowing gz compression to parquet will be a bigger win for space, and fully transparent to the end user.
To be honest, I think this one is a low priority; I think it can wait on a larger issue, which is whether there should a non-default, library-wide option to drop indexes. https://github.com/massivetexts/htrc-feature-reader/issues/20. Since that's a purely performance optimization, I think it can wait.

I imagine the argument would ultimately resolve to the 'write' method of the ParquetFileHandler. I think the most sensible way to handle is to have a generic `write` or `save` attribute on `Volume` that consumes an id resolver (or kwargs that can build one) and writes out from the current resolver to the new one.

This isn't implemented yet, but will build on this unit test from the ids notebook:

```python
def copy_between_resolvers(id, resolver1, resolver2):
    input = Volume(id, id_resolver=resolver1)
    output = Volume(id, id_resolver=resolver2, mode = 'wb')
    output.write(input)
```

```python
with tempfile.TemporaryDirectory() as first_new_dir:
    with tempfile.TemporaryDirectory() as second_new_dir:
        resolver1 = htrc_features.resolvers.LocalResolver(dir = Path(project_root, "tests", "data"), format = "json", compression = "bz2")
        resolver2 = htrc_features.resolvers.PairtreeResolver(dir = first_new_dir,  format = "json", compression = "gz")
        resolver3 = htrc_features.resolvers.LocalResolver(dir = second_new_dir, format = "parquet", compression = "snappy")
        
        copy_between_resolvers("aeu.ark:/13960/t1rf63t52", resolver1, resolver2)
        copy_between_resolvers("aeu.ark:/13960/t1rf63t52", resolver2, resolver3)
        
        all_files = []
        for loc, dir, files in os.walk(first_new_dir):
            for file in files:
                all_files.append(os.path.join(loc, file))
                
        assert(len(all_files) == 1)
        assert(all_files[0].endswith("aeu/pairtree_root/ar/k+/=1/39/60/=t/1r/f6/3t/52/ark+=13960=t1rf63t52/aeu.ark+=13960=t1rf63t52.json.gz"))
        
        # Our test assertion ensures that the data has made it all the way through.
        assert(Volume("aeu.ark:/13960/t1rf63t52", id_resolver = resolver3).tokenlist()['count'].sum() == 97691)
```

**Comment by bmschmidt** [comment]
Ah, I see, you've been fixing up the save_parquet method to do this at the same time. I can port those over later. I think we decided to move all 'write' logic to the file handlers from the base Volume class, right?
